### ISSUE TITLE
Achievements not working with expanded GBC memory map

### ISSUE THREAD

**Issue by rzumer** [libretro/RetroArch#8582]
## Description

Recently, Gambatte, SameBoy and Gearboy merged PRs to expand the memory map for the GBC system. The Gambatte core has an interface in the [RALibretro](https://github.com/RetroAchievements/RALibretro) frontend, and with it we can see that the expanded memory (0x10000 to 0x16FFF) is mapped as expected. However, when loaded in RetroArch, triggers that need to access memory in this area do not fire when they should. The same is true of other cores. Therefore it seems like there is an issue with RetroArch's achievements implementation that is blocking access to the memory.

Below is the memory map for the latest nightly version of Gambatte. The region with index 004 is the newly-mapped region.

```
[INFO]    ndx flags  ptr              offset   start    select   disconn  len      addrspace
[INFO]    001 M1A1bc 0000000012EEE040 00000000 0000C000 FFFFF000 00000000 00001000
[INFO]    002 M1A1bc 0000000012EEF040 00000000 0000D000 FFFFF000 00000000 00001000
[INFO]    003 M1A1bc 000000001351B430 00000000 0000FF80 FFFFFF80 00000000 00000080
[INFO]    004 M1A1bc 0000000012EEF040 00000000 00010000 00000000 00000000 00007000
[INFO]    005 M1A1bc 0000000012EEC040 00000000 0000A000 FFFFE000 00000000 00002000
[INFO]    006 M1A1bc 0000000012EE8040 00000000 00008000 00000000 00000000 00002000
[INFO]    007 M1A1bc 000000001351B2B0 00000000 0000FE00 00000000 00000000 000000A0
[INFO]    008 M1A1bC 0000000012AE8040 00000000 00000000 00000000 00000000 00004000
[INFO]    009 M1A1bC 0000000012AEC040 00000000 00004000 00000000 00000000 00004000
```

And below is how it is mapped in the source.

```
   struct retro_memory_descriptor descs[] =
   {
      { mainram, gb.rambank0_ptr(),  0, 0xC000,               0, 0, 0x1000,                        NULL },
      { mainram, gb.bankedram_ptr(), 0, 0xD000,               0, 0, 0x1000,                        NULL },
      { mainram, gb.zeropage_ptr(),  0, 0xFF80,               0, 0, 0x0080,                        NULL },
      { mainram, gb.rambank1_ptr(),  0, 0x10000,              0, 0, gb.isCgb() ? 0x7000 : 0,       NULL },
      {       0, gb.savedata_ptr(),  0, 0xA000, (size_t)~0x1FFF, 0, sramlen,                       NULL },
      {       0, gb.vram_ptr(),      0, 0x8000,               0, 0, 0x2000,                        NULL },
      {       0, gb.oamram_ptr(),    0, 0xFE00,               0, 0, 0x00A0,                        NULL },
      {     rom, gb.rombank0_ptr(),  0, 0x0000,               0, 0, 0x4000,                        NULL },
      {     rom, gb.rombank1_ptr(),  0, 0x4000,               0, 0, 0x4000,                        NULL },
   };
   
   struct retro_memory_map mmaps =
   {
      descs,
      sizeof(descs) / sizeof(descs[0]) - (sramlen == 0)
   };
```

### Steps to reproduce the bug

0. Get the latest nightly versions of RetroArch and Gambatte
1. Set up an account in the Achievements menu
2. Load "Harry Potter and the Sorcerer's Stone" and make sure that achievements are detected
3. Start a new game and get a wand (this takes a few seconds)

In RALibretro, with the latest 32-bit nightly build of Gambatte, an achievement should unlock. In RetroArch, it does not.

To test again after unlocking, log into the [RetroAchievements](https://retroachievements.org/) website, go to the user settings, and reset the unlock at the bottom of the page. Then unload and reload the game.

### Environment information

- OS: Windows (both 32- and 64-bit tested)

**Comment by rzumer** [comment]
I found the issue: `retro_get_memory_size()` is variable based on the system (GB or GBC), but apparently RetroArch does not refresh this value on load, so it is always cut off as the default system type is GB. This must be corrected at the core level, so closing.

**Comment by rzumer** [comment]
Actually, I will reopen, because I think changing making changes at the core level based on the assumption that RetroArch won't have an up-to-date memory info is not ideal. Since some cores don't allocate memory for both systems in the same location, it's a considerable amount of work needed to get around an issue with the achievements implementation.

The memory is fetched before even hashing/identifying the file being loaded. So most likely at the point where the memory info is collected in the achievements module, the system is not yet initialized, and this is why the wrong information is retrieved.

https://github.com/libretro/RetroArch/blob/master/cheevos/cheevos.c#L2644

**Comment by retro-wertz** [comment]
here's what i get

```
[INFO] [CHEEVOS]: address 1556B set to descriptor 17 at offset 0
[INFO] [CHEEVOS]: address 15567 set to descriptor 17 at offset 0
[INFO] [CHEEVOS]: address 15568 set to descriptor 17 at offset 0
[INFO] [CHEEVOS]: address 15569 set to descriptor 17 at offset 0
[INFO] [CHEEVOS]: address 1556A set to descriptor 17 at offset 0
[INFO] [CHEEVOS]: address 1556B set to descriptor 17 at offset 0
[INFO] [CHEEVOS]: address 1556C set to descriptor 17 at offset 0
[INFO] [CHEEVOS]: address 1556D set to descriptor 17 at offset 0
[INFO] [CHEEVOS]: address 1556E set to descriptor 17 at offset 0
```
offset 0, is this achievement even working?

**Comment by retro-wertz** [comment]
any update on this? after this has been merged, https://github.com/RetroAchievements/RAIntegration/pull/366, i cant get the achievement to work anymore. even following the new mapping (bank 0 @ $C000 (fixed bank), bank 1 @D000, bank 2-7 @ $10000)

**Comment by rzumer** [comment]
Still a problem with rcheevos enabled.

**Comment by rzumer** [comment]
My suspicious is that the root cause is a reliance on *both* memory descriptors and `retro_get_memory_data()` and `retro_get_memory_size()` functions to determine what memory regions are allowed to be monitored. These two functions can't be used to describe noncontiguous regions of RAM. Even when hardcoding the size of the RAM in the Gambatte core the banked RAM seems to be considered as "out of range" and does not get monitored. Note that it *does* work on another libretro frontend with achievement support (RAlibretro), using the exact same build for the core.

I am not planning on doing the work to fix this issue, but hopefully this gives anyone willing to do it all the information needed.

**Comment by rzumer** [comment]
Tested different configurations for setting up memory descriptors in `gambatte-libretro`, and took a look at the RA code. The address parsed for this test is set to a descriptor at offset 0, which is incorrect. With only the relevant memory region passed as a descriptor, it seems to unlock correctly despite the wrong offset being reported. With the rest of the memory map configured it fails completely. So it seems clear that virtual address -> memory descriptor + offset translation is bugged.

Testing with the RAlibretro frontend there is no issue with the same build of the core.

I deleted my previous message because I reached a wrong conclusion.

I don't intend on working on this issue further, but this should be enough information to assert that the issue is specific to RetroArch and comes from somewhere in the `cheevos-new` folder (note this is not a regression, the issue was there before the new implementation as well).
